sql/src/main/java/io/crate/metadata/Schemas.java

```diff
@@ -102,20 +105,113 @@ public Schemas(Settings settings,
         this.defaultTemplateService = new DefaultTemplateService(settings, clusterService);
     }
 
-    @Nullable
-    public TableInfo getTableInfoOrNull(RelationName relationName, Operation operation) {
-        SchemaInfo schemaInfo = schemas.get(relationName.schema());
-        if (schemaInfo == null) {
-            return null;
+    public TableInfo resolveTableInfo(QualifiedName ident, Operation operation, SearchPath searchPath) {
+        String identSchema = schemaName(ident);
+        String tableName = relationName(ident);
+
+        SchemaInfo schemaInfo = null;
+        TableInfo tableInfo = null;
+        if (identSchema == null) {
+            for (String pathSchema : searchPath) {
+                schemaInfo = schemas.get(pathSchema);
+                if (schemaInfo != null) {
+                    tableInfo = schemaInfo.getTableInfo(tableName);
+                    if (tableInfo != null) {
+                        break;
+                    }
+                }
+            }
+        } else {
+            schemaInfo = schemas.get(identSchema);
+            if (schemaInfo == null) {
+                throw new SchemaUnknownException(identSchema);
+            } else {
+                tableInfo = schemaInfo.getTableInfo(tableName);
+            }
         }
-        TableInfo tableInfo = schemaInfo.getTableInfo(relationName.name());
-        if (tableInfo == null) {
-            return null;
+
+        if (schemaInfo == null || tableInfo == null) {
```

Comment:
minor: afaik the either the tableInfo == null or schemaInfo != null, having schemaInfo == null but tableInfo not cannot happen afaik. (so this could be simplified by `if (tableInfo == null)`